File: 45-jump-game-ii/jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:

        memo = [10001] * len(nums)

        return self.jumpHelper(nums, 0, memo)
        
    def jumpHelper(self, nums, index, memo):
        if index >= len(nums) - 1:
            return 0
       
        if memo[index] != 10001:
            return memo[index]

        possible_jump = nums[index]

        for jump in range(1, possible_jump+1):
            memo[index] = min(memo[index], self.jumpHelper(nums, index + jump, memo) + 1)
        
        return memo[index]
        

File: 45-jump-game-ii/jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        if (nums.size() <= 1) return 0;

        int jumps = 0;
        int current_reach = 0; 
        int max_reach = 0;  

        for (int i = 0; i < nums.size() - 1; ++i) {
            max_reach = max(max_reach, i + nums[i]);

            if (i == current_reach) { 
                jumps++;
                current_reach = max_reach; 
                if (current_reach >= nums.size() - 1) return jumps; 
            }
        }
        return jumps;
    }
};